File: sqlite3_web/lib/src/channel.dart
import 'dart:async';
import 'dart:js_interop';
import 'dart:math';

import 'package:web/web.dart' hide Request, Response, Notification;

import 'locks.dart';
import 'protocol.dart';
import 'types.dart';

const _disconnectMessage = '_disconnect';
final Random _random = Random();

@JS()
@anonymous
extension type WebEndpoint._(JSObject _) implements JSObject {
  external MessagePort get port;
  external String? get lockName;

  external factory WebEndpoint({
    required MessagePort port,
    required String? lockName,
  });

  ConnectableChannel connect() {
    return ConnectableChannel._(port, lockName, null);
  }
}

final class ConnectableChannel {
  final MessagePort localPort;
  final String? lockName;
  final HeldLock? lock;
  EventTarget? injectErrors;

  ConnectableChannel._(this.localPort, this.lockName, this.lock);
}

Future<(WebEndpoint, ConnectableChannel)> createChannel() async {
  final webChannel = MessageChannel();
  final locks = WebLocks.instance;

  // For servers, it's useful to know when a connection is closed, so that it
  // can release locks or otherwise clean up resources occupied by the
  // connection. This is kind of tricky for message ports, because they have
  // no "closed" event.
  // As a workaround, we're making one side take a unique lock and tell the
  // other endpoint about it. When that side closes for any reason, even just
  // because the tab is improperly closed, the other end will be able to take
  // the lock and thus know that the connection has been closed.
  HeldLock? lock;
  String? lockName;
  if (locks != null) {
    lockName = _randomLockName();
    lock = await locks.request(lockName);
  }

  final channel = ConnectableChannel._(webChannel.port2, lockName, lock);
  return (WebEndpoint(port: webChannel.port1, lockName: lockName), channel);
}

String _randomLockName() {
  final buffer = StringBuffer('channel-close-');
  for (var i = 0; i < 16; i++) {
    const charCodeSmallA = 97;
    buffer.writeCharCode(charCodeSmallA + _random.nextInt(26));
  }

  return buffer.toString();
}

abstract base class ProtocolChannel extends RequestHandler {
  final MessagePort _port;
  final Completer<void> _closed = Completer();
  StreamSubscription<void>? _incomingMessagesSubscription;
  StreamSubscription<void>? _errorSubscription;

  var _nextRequestId = 0;
  final Map<int, Completer<Response>> _responses = {};

  /// Requests that are currently being handled, identified by their id. This
  /// allows aborting them.
  final Map<int, AbortController> _handlingRequests = {};

  ProtocolChannel(ConnectableChannel connectable)
    : _port = connectable.localPort {
    _port.start();

    _incomingMessagesSubscription = EventStreamProviders.messageEvent
        .forTarget(_port)
        .listen((event) {
          final data = event.data;
          if (data.equals(_disconnectMessage.toJS).toDart) {
            _markClosed();
            return;
          }

          _handleIncoming(event.data as Message);
        });

    if (connectable.injectErrors case final injectErrors?) {
      _errorSubscription = EventStreamProviders.errorEvent
          .forTarget(injectErrors)
          .listen((event) {
            final error = (event as ErrorEvent).error;
            _markClosed(error);
          });
    }

    final lockName = connectable.lockName;
    if (connectable.lock == null && lockName != null) {
      // Once this side is able to acquire the lock, the connection is closed.
      WebLocks.instance!.request(lockName).then((lock) {
        _markClosed();
        lock.release();
      });
    }
  }

  Future<void> get closed => _closed.future;

  void _send(Message message) {
    message.sendToPort(_port);
  }

  /// Handle an incoming message from the client.
  void _handleIncoming(Message message) async {
    dispatchMessage(
      message,
      whenResponse: (response) {
        _responses.remove(response.requestId)?.complete(response);
      },
      whenRequest: (request) async {
        Response response;

        final requestId = request.requestId;
        final abortController = _handlingRequests[requestId] =
            AbortController();

        try {
          response = await dispatchRequest(request, abortController.signal);
        } catch (e, s) {
          if (e is! AbortException) {
            console.error('Error in worker: ${e.toString()}'.toJS);
            console.error('Original trace: $s'.toJS);
          }

          response = ErrorResponseUtils.wrapException(requestId, e);
        } finally {
          _handlingRequests.remove(requestId);
        }

        _send(response);
      },
      whenNotification: handleNotification,
      whenAbortRequest: (abort) {
        if (_handlingRequests.remove(abort.requestId) case final token?) {
          token.abort();
        }
      },
    );
  }

  /// Sends a request to the other end and expects a response of the
  /// [expectedType].
  ///
  /// The returned future completes with an error if the type doesn't match
  /// what's expected (for instance because an [ErrorResponse]) is sent instead.
  /// It also completes with an error if the channel gets closed in the
  /// meantime.
  ///
  /// If [abortTrigger] is given and completes before this request is completed,
  /// a request to cancel the request is sent to the remote.
  Future<Res> sendRequest<Res extends Response>(
    Request request,
    MessageType<Res> expectedType, {
    Future<void>? abortTrigger,
  }) async {
    if (_closed.isCompleted) {
      throw ChannelClosedException._();
    }

    final id = _nextRequestId++;
    final completer = _responses[id] = Completer.sync();

    _send(request..requestId = id);
    var hasResponse = false;

    if (abortTrigger != null) {
      abortTrigger.whenComplete(() {
        if (!hasResponse) {
          _send(newAbortRequest(requestId: id));
        }
      });
    }

    final response = await completer.future;
    hasResponse = true;
    if (response.type == expectedType.name) {
      return response as Res;
    } else {
      throw response.interpretAsError();
    }
  }

  void sendNotification(Notification notification) {
    _send(notification);
  }

  void handleNotification(Notification notification);

  Future<void> close([Object? error]) {
    _markClosed(error);
    return closed;
  }

  void _markClosed([Object? error]) {
    if (_closed.isCompleted) return;

    _port.postMessage(_disconnectMessage.toJS);
    _incomingMessagesSubscription?.cancel();
    _errorSubscription?.cancel();

    for (final response in _responses.values) {
      response.completeError(ChannelClosedException._(error));
    }
    _responses.clear();

    _closed.complete();
  }
}

/// An exception thrown when a request is sent over a closed channel to a
/// worker.
final class ChannelClosedException implements Exception {
  /// The original error causing the channel to be closed.
  final Object? closeReason;

  ChannelClosedException._([this.closeReason]);

  @override
  String toString() {
    return 'Channel to database worker is closed: $closeReason';
  }
}

File: sqlite3_web/extra/js-package/src/channel.ts
import {
  AbortRequest,
  extractTransferrable,
  Message,
  Request,
  Response,
  Notification,
  dispatchMessage,
  typeAbortRequest,
} from "./generated_protocol.js";
import {
  interpretResponseAsError,
  ReleaseLock,
  requestNavigatorLock,
  serializeError,
} from "./utils.js";

const disconnectMessage = "_disconnect";

export interface WebEndpoint {
  port: MessagePort;
  lockName: string | null;
}

export function randomLockName() {
  return `sqlite3-web-${crypto.randomUUID()}`;
}

export async function createChannel(
  options: Partial<ProtocolChannelOptions> = {},
): Promise<[WebEndpoint, ProtocolChannelOptions]> {
  const { port1, port2 } = new MessageChannel();
  // Request a random lock. We include its name in WebEndpoint. Since tabs return locks
  // when they're closed, that allows clients to reliably detect dead remotes.
  const lockName = randomLockName();
  const lock = await requestNavigatorLock(lockName);

  return [
    { port: port2, lockName },
    {
      ...options,
      _internal_port: port1,
      _internal_lockName: lockName,
      _internal_lock: lock,
    },
  ];
}

export interface ProtocolChannelOptions {
  _internal_port: MessagePort;
  _internal_lockName: string | null;
  _internal_lock?: ReleaseLock;
  _internal_errors?: EventTarget;
}

export abstract class ProtocolChannel {
  #port: MessagePort;
  #eventListener: (event: MessageEvent) => void;
  #heldLock?: ReleaseLock;

  #nextRequestId = 0;
  #outstandingRequests = new Map<
    number,
    { ok: (e: Response) => void; e: (e: unknown) => void }
  >();
  #closed = false;
  #closedPromise: Promise<void>;
  #markClosed!: () => void;

  constructor({
    _internal_port: port,
    _internal_lockName: lockName,
    _internal_lock: lock,
    _internal_errors: errors,
  }: ProtocolChannelOptions) {
    this.#port = port;
    this.#closedPromise = new Promise((resolve) => {
      this.#markClosed = () => {
        this.#closed = true;
        resolve();
      };
    });

    if (lock == null && lockName != null) {
      // Once this side is able to acquire the lock, the connection is closed.
      navigator.locks.request(lockName, () => {
        this.#markRemoteClosed();
      });
    }

    errors?.addEventListener(
      "error",
      (event) => {
        this.#markRemoteClosed((event as ErrorEvent).error);
      },
      { once: true },
    );

    this.#port.start();

    this.#eventListener = (event) => {
      if (event.data === disconnectMessage) {
        this.#markRemoteClosed();
      } else {
        this.#handleMessage(event.data);
      }
    };
    this.#port.addEventListener("message", this.#eventListener);
  }

  public get _internal_closed(): Promise<void> {
    return this.#closedPromise;
  }

  #handleMessage(message: Message) {
    const self = this;

    dispatchMessage(message, {
      _internal_whenAbortRequest() {
        // We don't currently support aborting requests handled on this side.
      },
      _internal_whenNotification(notification) {
        self._internal_handleNotification(notification);
      },
      _internal_whenResponse: function (message: Response) {
        const entry = self.#outstandingRequests.get(message.i);
        self.#outstandingRequests.delete(message.i);
        entry?.ok(message);
      },
      async _internal_whenRequest(message: Request) {
        let response: Response;
        try {
          response = await self._internal_serveRequest(message);
        } catch (e) {
          response = serializeError(message.i, "Error in JS client", e);
        }

        self.#port.postMessage(response, extractTransferrable(response));
      },
    });
  }

  #markRemoteClosed(cause?: unknown) {
    this.#closeLocally(cause);
  }

  #closeLocally(cause?: unknown) {
    this.#markClosed();
    this.#outstandingRequests.forEach((value) => {
      value.e(new Error(`Channel closed before receiving response: ${cause}`));
    });

    this.#port.postMessage(disconnectMessage);
    this.#port.removeEventListener("message", this.#eventListener);
    this.#port.close();

    this.#heldLock?.();
  }

  _internal_close() {
    return this.#closeLocally();
  }

  abstract _internal_serveRequest(
    request: Request,
  ): Promise<Response> | Response;

  abstract _internal_handleNotification(notification: Notification): void;

  _internal_sendNotification(notification: Notification) {
    this.#port.postMessage(notification, extractTransferrable(notification));
  }

  async _internal_sendRequest<Req extends Request, Res extends Response>(
    request: Omit<Req, "i">,
    expectedType: Res["t"],
    abort?: AbortSignal,
  ): Promise<Res> {
    abort?.throwIfAborted();

    if (this.#closed) {
      throw new Error("Channel closed");
    }

    let hasResponse = false;
    const response = await new Promise<Response>((ok, e) => {
      const id = this.#nextRequestId++;
      this.#outstandingRequests.set(id, { ok, e });

      (request as Req).i = id;
      this.#port.postMessage(request, extractTransferrable(request as Req));

      if (abort) {
        abort.addEventListener("abort", () => {
          if (!hasResponse && !this.#closed) {
            this.#port.postMessage({
              t: typeAbortRequest,
              i: id,
            } satisfies AbortRequest);
          }
        });
      }
    });
    hasResponse = true;

    if (response.t === expectedType) {
      return response as Res;
    } else {
      throw interpretResponseAsError(response);
    }
  }
}
